plugins/techdocs-addons-test-utils/src/test-utils.tsx

```diff
@@ -0,0 +1,240 @@
+    return this;
+  }
+
+  withEntity(entity: RecursivePartial<TechDocsEntityMetadata>) {
+    this.options.entity = entity;
+    return this;
+  }
+
+  atPath(path: string) {
+    this.options.path = path;
+    return this;
+  }
+
+  build() {
+    const apis = [
+      [techdocsApiRef, techdocsApi],
+      [techdocsStorageApiRef, techdocsStorageApi],
+      [searchApiRef, searchApi],
+      ...this.options.apis,
+    ];
+
+    const entityName = {
+      namespace:
+        this.options.entity?.metadata?.namespace ||
+        defaultEntity.metadata.namespace,
+      kind: this.options.entity?.kind || defaultEntity.kind,
+      name: this.options.entity?.metadata?.name || defaultEntity.metadata.name,
+    };
+
+    techdocsApi.getTechDocsMetadata.mockReturnValue(
+      this.options.metadata || { ...defaultMetadata },
+    );
+    techdocsApi.getEntityMetadata.mockResolvedValue(
+      this.options.entity || { ...defaultEntity },
+    );
+
+    techdocsStorageApi.syncEntityDocs.mockResolvedValue('cached');
+    techdocsStorageApi.getApiOrigin.mockResolvedValue(
+      'https://backstage.example.com/api/techdocs',
+    );
+
+    techdocsStorageApi.getEntityDocs.mockResolvedValue(
+      renderToStaticMarkup(this.options.dom || defaultDom),
+    );
+
+    const TechDocsAddonsPage = () => {
+      return (
+        <TestApiProvider apis={apis}>
+          <FlatRoutes>
+            <Route
+              path="/docs/:namespace/:kind/:name/*"
+              element={<TechDocsReaderPage />}
+            >
+              <TechDocsAddons>
+                {this.addons.map((addon, index) => (
+                  <Fragment key={index}>{addon}</Fragment>
+                ))}
+              </TechDocsAddons>
+            </Route>
+          </FlatRoutes>
+        </TestApiProvider>
+      );
+    };
+
+    return wrapInTestApp(<TechDocsAddonsPage />, {
+      routeEntries: [
+        `/docs/${entityName.namespace}/${entityName.kind}/${entityName.name}/${this.options.path}`,
+      ],
+      mountedRoutes: {
+        '/docs': techdocsPlugin.routes.root,
+        '/docs/:namespace/:kind/:name/*': techdocsPlugin.routes.docRoot,
+        '/catalog/:namespace/:kind/:name': catalogPlugin.routes.catalogEntity,
+      },
+    });
+  }
+
+  // Components using useEffect to perform an asynchronous action (such as fetch) must be rendered within an async
+  // act call to properly get the final state, even with mocked responses. This utility method makes the signature a bit
+  // cleaner, since act doesn't return the result of the evaluated function.
+  // https://github.com/testing-library/react-testing-library/issues/281
+  // https://github.com/facebook/react/pull/14853
+  async renderWithEffects(): Promise<
+    typeof screen & { shadowRoot: ShadowRoot | null }
+  > {
+    await act(async () => {
+      render(this.build());
+    });
+
+    const shadowHost = screen.getByTestId('techdocs-native-shadowroot');
+
+    return {
+      ...screen,
```

Comment:
How come this is the screen rather than the result of the `render` call?

Reply:
This is because we are using a screen that has support for Shadow DOM testing utilities:
https://github.com/testing-library/dom-testing-library/issues/742#issuecomment-674987855

Reply:
Confused about where the source code for that package is 😅 . Alright, that explains why it's a bit different, still kinda would expect that library to have a `render` method exported though, but if it doesn't then it doesn't x)

Reply:
I agree... But it doesn't export a render, so we used the screen.